File: sjq-leetcode-C++/rotate-array.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        // 调转三次
        if(nums.size()<2) return;
        k %= nums.size();
        reverse(nums, 0, nums.size()-1);
        reverse(nums, 0, k-1);
        reverse(nums, k, nums.size()-1);
    }
    
    void reverse(vector<int>& nums, int l, int r){
        while(l < r)
            // 交换前后两个元素
            swap(nums[l++], nums[r--]);
    }
};

File: zs-leetcode-go/RotateArray/RotateArray.go
package RotateArray

func reverse(nums []int) {
	for start, end := 0, len(nums)-1; start < end; start, end = start+1, end-1 {
		nums[start], nums[end] = nums[end], nums[start]
	}
}

func rotate(nums []int, k int) {
	length := len(nums)
	if length == 0 || k == 0 {
		return
	}
	k %= length
	reverse(nums)
	reverse(nums[:k])
	reverse(nums[k:])
}
